### PR TITLE
Merge MySQL 5.7.9 (GA) semantics into MySQL57Platform

### PR DESCRIPTION
**Pull request by deeky666**
This is adopted from #2266 and #2455 and contains additional fixes, adjustments and upgrade notice.

I made the following decisions about this whole topic:

- Merge the changes into `MySQL57Platform` as suggested by @beberlei because the first GA version of the `5.7` series is `5.7.9`
- Accept possible minor BC breaks when upgrading the schema (tracked in `UPGRADE.md`)
- Do not care about invalid values previously store in `json_array` type columns (they would be invalid with `2.5` anyways so what's the point?)
- Did I forget anything? Please let me know.

### CODE REVIEW

File: lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php
Comment (phansys) [inline lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php]: Please, use Yoda condition.

File: tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php
Comment (phansys) [inline tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php]: Support for PHP < 7 has been dropped at `master`. Could we use `[]` instead?

**Review by phansys — CHANGES_REQUESTED** [review]

File: tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php
Comment (deeky666) [inline tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php]: @phansys is that really important here? We would need to fix all of that stuff in one PR anyways.

File: lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php
Comment (deeky666) [inline lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php]: Fixed

**Comment by rh389** [comment]
I'm glad to see some movement on this. Less enthused that it's taken you 14 months to rip off the work of other contributors and ignore carefully investigated solutions to BC breaks. 👏

**Comment by Ocramius** [comment]
> rip off the work

You are always welcome to rip off this piece of work by forking the repository and then maintaining it yourself. Seriously, refrain from posting stuff like this: we do what we can in our own free time, so either be happy with that or pay for software maintenance in the closed source world.

**Comment by deeky666** [comment]
@rh389 I did not intend to rip anyone's work off. Originally I wanted to use either of the previous PRs getting merged but both of them were lacking some things. #2266 need merge into `MySQL57Platform`, #2455 needed keywords, tests, documentation etc. And both of them did not have the upgrade entry. As many people like you want to get this finally going, I decided to take a "fast" approach by using what was already contributed (btw see the `@author` tag in `MySQL57Platform`) and put the missing pieces in. So much to ripping off.
As to why this took 14 months, as you might have noticed, I took a year off of OS development last year (more or less) and I apologize if that might have been frustrating to some, but we all do this in our free time and have personal lives beside Doctrine.

**Comment by rh389** [comment]
Sorry @deeky666, I was out of order. It was frustrating to work on this issue without a response but that's OS sometimes, not your fault. Apologies again, thanks for you efforts and looking forward to seeing this land.

File: UPGRADE.md
Comment (rh389) [inline UPGRADE.md]: I'm not sure this is actually true - the `Comparator` doesn't inspect the actual column type so won't pick up the change (unless/until it has some other reason to alter that column). See https://github.com/doctrine/dbal/pull/2266#issuecomment-254499121

Comment (deeky666) [inline UPGRADE.md]: @rh389 good point. Considering that, I can remove the upgrade notice again I suppose unless there are other reasons why BC could not be maintained here. Currently I can't think of any.

Comment (rh389) [inline UPGRADE.md]: Well, it will alter to a native `JSON` type later if the user changes their column definition some other way, which might be a bit unexpected (buggy?) but I guess technically not a BC break.

When the column is updated though it's actually quite likely to fail in practice because the `LONGTEXT` implementation used an empty string as a default value, which is invalid for native `JSON`. It's not just the case that it'll only fail if the user was doing something weird.

Not a blocker IMO - just making sure you're aware.

Comment (deeky666) [inline UPGRADE.md]: I see, there is nothing we can do about it if we want to move on. I'll remove the upgrade notice then.

**Comment by deeky666** [comment]
@Ocramius @phansys @rh389 this thing is ready from my point of view. Anything we forgot?

File: lib/Doctrine/DBAL/Platforms/Keywords/MySQL57Keywords.php
Comment (phansys) [inline lib/Doctrine/DBAL/Platforms/Keywords/MySQL57Keywords.php]: This should be placed after `'STARTING'` and before `'STRAIGHT_JOIN'` in order to keep the current sorting.

File: tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php
Comment (phansys) [inline tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php]: We should use `assertSame()` instead.

**Review by phansys — CHANGES_REQUESTED** [review]
Status: Needs work.

File: lib/Doctrine/DBAL/Platforms/Keywords/MySQL57Keywords.php
Comment (deeky666) [inline lib/Doctrine/DBAL/Platforms/Keywords/MySQL57Keywords.php]: Fixed.

File: tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php
Comment (deeky666) [inline tests/Doctrine/Tests/DBAL/Platforms/MySQL57PlatformTest.php]: Fixed.

**Review by jnevesjunior — APPROVED** [review]

**Comment by Omranic** [comment]
@deeky666 @Ocramius @phansys @rh389 Is this PR landing anytime soon?! 🤔  Thanks in advance

**Comment by deeky666** [comment]
@Omranic waiting for an approval and merge from one of the other core members.

**Comment by mpyw** [comment]
Pending for 27 days......

**Comment by Ocramius** [comment]
That is quite short

**Comment by kimhemsoe** [comment]
@deeky666 Looks good to me

**Comment by mpyw** [comment]
@kimhemsoe @deeky666 I'm also using the branch in my product and working fine 👍

**Review by Ocramius — APPROVED** [review]
LGTM :ship:

**Comment by mpyw** [comment]
@Ocramius Thank you ✨

**Comment by MartinVandersteen** [comment]
Hello guys, first of all thanks a lot for your work.
I have a small question though, I upgraded to PHP7.1 and dbal 2.6.1 especially to fix a bug that I'm currently experiencing. I'm just trying to create a json database field in one of my laravel migrations.

I still get this message though : 
```
[Doctrine\DBAL\DBALException]                                                                 
  Unknown database type json requested, Doctrine\DBAL\Platforms\MySqlPlatform may not support   
  it. 
``` 

Which makes me think that MySQL57Platform is never used or acknowledged? Any hints about where it could come from? Thanks in advance! Have a great week.

**Comment by Ocramius** [comment]
@MartinVandersteen please report a separate issue with a way to reproduce it.

**Comment by MartinVandersteen** [comment]
I don't think there is any way I could find a way to reproduce it, just asking if something has to be done at some point for Dbal to understand it has to use MySQL57 ? On what ground does it decide which one to use?

**Comment by Ocramius** [comment]
> just asking if something has to be done at some point for Dbal to understand it has to use MySQL57 

See https://github.com/doctrine/dbal/blob/399ab63ddf3db48a4befb7d4f6bcf699e7796d5f/lib/Doctrine/DBAL/Connection.php#L391

A test would:

 * connect to a DB
 * create the schema
 * try to introspect the schema and verify that the "json" type was correctly detected

That said, please report further findings on a new issue.

**Comment by hansbogert** [comment]
The Type matrix does not reflect this change. In fact it doesn't show the `json` type at all. (it does show the deprecated `json_array`)

**Pull request merged**
